model-checkers/src/main/java/de/learnlib/modelchecking/modelchecker/LTSminLTL.java

```diff
@@ -0,0 +1,413 @@
+/* Copyright (C) 2013-2017 TU Dortmund
+ * This file is part of LearnLib, http://www.learnlib.de/.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package de.learnlib.modelchecking.modelchecker;
+
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.file.CopyOption;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.StandardCopyOption;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.function.Function;
+
+import de.learnlib.api.modelchecking.counterexample.Lasso;
+import de.learnlib.api.modelchecking.modelchecker.ModelChecker;
+import de.learnlib.api.modelchecking.modelchecker.ModelCheckingException;
+import net.automatalib.automata.concepts.Output;
+import net.automatalib.parser.FSMParseException;
+import net.automatalib.parser.FSMParser;
+import net.automatalib.serialization.etf.writer.ETFWriter;
+import net.automatalib.ts.simple.SimpleDTS;
+
+/**
+ * An LTL model checker using LTSmin.
+ *
+ * Key in this implementation is that the LTSmin binaries are extracted from the JAR file and written to a temporary
+ * file (which is deleted when the JVM exits).
```

Comment:
Well something can be said for both imo. When shipping the binaries LearnLib remains quite portable. And I know from experience that users/students find it quite difficult to install LTSmin (even when it involves only setting the PATH variable). We don't have LTSmin installers.

The binaries I currently distribute with LearnLib (in this PR) can be downloaded from here: https://github.com/utwente-fmt/ltsmin/releases/tag/3.0. They are standalone and don't require extra dependencies.
So the developer only needs to copy a binary from the github release page. I could write some more Javadoc so that a developer knows where to get more recent versions of LTSmin?

Also, if we allow only user installed LTSmin binaries then I would have to write some version checks which I don't really know how to do in Java? How can LearnLib check someone exactly installed a specific LTSmin version? We won't have this problem if LearnLib comes with the correct LTSmin version.
Anyway, I am fine with removing LTSmin from the LearnLib distribution but that comes with a disadvantage too?

Reply:
It's nice that you have ootb binaries of LTSmin (makes it nice to setup in travis, too). Looking at the github tag, you provide MacOS and Linux binaries. What if someone actually compiles his own version of LTSmin for Windows? (Java certainly supports a variety of platforms.) Well, still can't use LearnLib, as only MacOS and Linux are packaged. What if you detect a memory leak and have to release version 3.0.1? Well, guess we're gonna have to roll out a new LearnLib release as well, if you want to check that big system with it ... you see, where I'm heading? I'm strongly in favor of handling LTSmin outside of LearnLib.

Regarding the configuration of the $PATH variable: You could add support for parsing an environment property via `LearnLibSettings` (e.g. `de.learnlib.external.ltsmin.bin`?) for detecting the binary (i.e., when the property is set, use it, otherwise fallback to what $PATH gives). This way, if it is too cumbersome to correctly setup the $PATH variable, people could still simply use `java -Dde.learnlib.external.ltsmin.bin=/path/to/ltsmin -jar learnlib.jar`.

Regarding sanity checks: Does LTSmin support a `-v` or `--version` flag? Basically, you can just call the binary and parse whatever output in generates, just like on a terminal. See e.g. the [DOT class](https://github.com/LearnLib/automatalib/blob/develop/visualization/dot-visualizer/src/main/java/net/automatalib/visualization/dot/DOT.java#L90) where we detect DOT support by simply checking if a call to the binary returns without any error. What do you need the version check for? Is it that old versions don't support e.g. the ETF format? If LTSmin has explanatory error messages (e.g. "Unkown file type: ETF"), I'd say it's also fine to just forward the error messages of the binary and not bother too much with version checking. The documentation (either JavaDoc or Maven Site) should however state, what the least expected version is.